File: Ilya-Dudnikov/term_2/block_4/task_2/bash/src/main/java/BashProject/shellcommand/commands/Cat.java
package BashProject.shellcommand.commands;

import BashProject.shellcommand.Command;

import java.io.*;
import java.nio.ByteBuffer;

public class Cat extends Command {
	private String executeCat(File file) {
		if (file != null && file.isDirectory()) {
			return "";
		}

		InputStream inputStream;
		if (file == null) {
			inputStream = System.in;
		} else {

			try {
				inputStream = new FileInputStream(file);
			} catch (FileNotFoundException e) {
				throw new RuntimeException(e);
			}
		}

		try {
			String result = new String(inputStream.readAllBytes());
			inputStream.close();
			return result;
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	@Override
	public ByteBuffer run(String ... arguments) {
		super.run(arguments);
		if (args.isEmpty()) {
			return ByteBuffer.wrap((executeCat(null) + System.lineSeparator()).getBytes());
		}

		StringBuilder result = new StringBuilder();
		for (int i = 0; i < args.size(); i++) {
			var fileName = args.get(i);
			File file = new File(fileName);

			if (!file.exists()) {
				result
						.append("cat: ")
						.append(fileName)
						.append(": No such file or directory")
						.append(System.lineSeparator());
				continue;
			}

			if (file.isDirectory()) {
				result
						.append("cat: ")
						.append(fileName)
						.append(": Is a directory");
			}

			result.append(executeCat(file)).append(System.lineSeparator());
		}

		return ByteBuffer.wrap(result.toString().getBytes());
	}
}

File: ShangareevRustam/Second term/Block 4/2. Simple bash interpreter/app-bash/src/main/kotlin/bashCommands/Cat.kt
package bashCommands

import simpleBashInterpreter.textUtilities.TextUtilities.trimQuotes
import java.io.BufferedReader
import java.io.File
import java.io.FileReader

object Cat : BashCommand() {

	override fun run(arg: String, input: String, arguments: MutableMap<String, String>): String {
		if (arg.split(" ").none { it != "" && it != "cat" }) return input
		val interpolatedArg = trimQuotes(arg.slice(arg.indexOf("cat") + 4 until arg.length).trim())

		val file = File(interpolatedArg)
		if (!file.exists()) return "cat: $interpolatedArg: No such file or directory"
		if (file.isDirectory) return "cat: $interpolatedArg: Is a directory"

		val result = StringBuilder()
		try {
			val bufferedReader = BufferedReader(FileReader(file))
			while (true) {
				val line = bufferedReader.readLine()
				result.append(if (line == null) break else "$line\n")
			}
			bufferedReader.close()
		} catch (e: Exception) {
			return "cat: $interpolatedArg: An error occurred while reading the file"
		}
		return result.toString()
	}

	override fun isValid(command: String) =
		command.split(" ").any { it != "" } &&
				command.split(" ").first { it != "" } == "cat"

}